src/ripple/app/main/Application.cpp

```diff
@@ -1110,7 +1064,30 @@ class ApplicationImp : public Application, public RootStoppable, public BasicApp
                 return validators().trustedPublisher(pubKey);
             });
 
-        stopped();
+        // The order of these stop calls is delicate.
+        // Re-ordering them risks undefined behavior.
+        m_shaMapStore->stop();
+        m_jobQueue->stop();
```

Comment:
You're mistaken. This is the consequence of `Stoppable` being so misleading. `Stoppable` gave the impression that children depend on their parents, parents would stop after their children, and because `JobQueue` was so high in the hierarchy, it would stop late in the sequence.

In fact, `JobQueue` was the FIRST `Stoppable` stopped, and it was [stopped](https://github.com/ripple/rippled/blob/c11037fd2764f6fea8e0d184b0b7aa0daa81136f/src/ripple/core/impl/Stoppable.cpp#L214) in `RootStoppable::stop`. Once the `jobCounter` (which is a `ClosureCounter`) is joined, no more jobs will be [created](https://github.com/ripple/rippled/blob/develop/src/ripple/core/JobQueue.h#L168-L173) (all "wrappers" will be [empty](https://github.com/ripple/rippled/blob/c11037fd2764f6fea8e0d184b0b7aa0daa81136f/src/ripple/core/ClosureCounter.h#L178-L187)), and the call to `ClosureCounter::join` [blocks](https://github.com/ripple/rippled/blob/c11037fd2764f6fea8e0d184b0b7aa0daa81136f/src/ripple/core/ClosureCounter.h#L156-L166) until they have all finished executing. There are no jobs and there will not be any more jobs before the first `Stoppable::onStop` is even entered.

Reply:
Luckily, it appears that `GRPCServer` is prepared to [handle](https://github.com/ripple/rippled/blob/c11037fd2764f6fea8e0d184b0b7aa0daa81136f/src/ripple/app/main/GRPCServer.cpp#L130-L135) the case where `JobQueue` stops before it.

Reply:
You're right, I let myself get mislead. `onStop()` is called parent down, though `m_stopped` seems to be set to `true` child up.